_RFCs/0004-testing-pull-requests.md

```diff
@@ -0,0 +1,47 @@
+
+Part of what makes pull requests to core plans linger open for so long is that, although many people are comfortable with reviewing the code, many are hesitant to merge, release, and promote the package for fear of breaking it or other packages dependent on it.  A future automated testing framework (currently in development) will automate much of this testing, but before we can automate testing something, we need to know how to test it manually.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+When someone opens a pull request to the core plans repo, they must include steps to test the change they are making (or the functionality of a new plan). The intent is to have a way to test any new functionality and, as much as possible, prevent breaking of existing functionality. Providing these testing steps (and sample apps when useful) is the responsibility of the person opening the pull request. However, should the person opening the pull request need help with establishing testing steps, they are most welcome to ask for it in the pull request.  A pull request must not be approved until testing steps are provided in the pull request.
+
+At least one core maintainer must review the pull request's code and, upon approval, either that same maintainer or another maintainer must follow the manual testing steps locally. This manual testing can be done either before or after the merge (when the package is on the unstable channel in Builder), but must be done before the package is promoted to stable. It should be noted that it is a bad practice to leave a package in unstable more than a day without testing it for promotion to stable (ideally, it should be tested as soon as it builds successfully and is put in the unstable channel).
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Here are some examples of manually testing pull requests:
+
+* [Reviewing a binary package](https://forums.habitat.sh/t/reviewing-core-plans-example-binaries/449)
+* [Reviewing changes to core/scaffolding-node and core/scaffolding-ruby](https://forums.habitat.sh/t/reviewing-core-plans-example-scaffoldings/450)
+* [General guidelines for reviewing core plans](https://forums.habitat.sh/t/reviewing-core-plans/434)
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+This does put more of a burden on the person who is opening the pull request - to not only submit the request, but also include steps for testing it.
+
+# Rationale and alternatives
+[alternatives]: #alternatives
+
+There is an automated testing framework in the works, which will be very helpful with this. That said, as of the time of writing this pull request, that framework is not yet ready. Additionally, even when that framework is ready, we still need to know how to manually test a plan before we can automate those tests. The best person to guide how to test a change is the person who made that change.
```

Comment:
For packages that include automated tests (as many gnome projects do for example), implementing the `do_check` build hook seems like the automated framework we'd need for binary packages.

Reply:
agree with @rsertelon.  We should ensure there is a valid `do_check` in core plans.  However there should be another form of functional testing when it is integrated with `hab-sup`.  Not sure what that framework is, but it should at least address the following:

  - can it start up?
  - can you configure the thing as expected?
  - can you run some health check?

Reply:
@eeyun and I have been exploring using Concourse for just that scenario :)

Reply:
Yep, in fact I've already written the pipeline to do this, we're just finishing up the deployment of the services and debugging any of the weird behavior we see in acceptance! So we should have examples asap!